sympy/sets/sets.py

```diff
@@ -1202,28 +1202,32 @@ def reduce(args):
 
         # ===== Pair-wise Rules =====
         # Here we depend on rules built into the constituent sets
-        args = set(args)
+        args = FiniteSet(*args)
```

Comment:
Sometimes the undocumented behavior depends on random seed which can lead to crazy bugs which sometimes appear in the code and sometime they don't. Maybe the [problem you mentioned above](https://github.com/sympy/sympy/pull/11188#issuecomment-229264778) arises because of that.